### PR TITLE
Testing robustness of bindMount

### PR DESCRIPTION
**Pull request by thenor**
Add more test cases to verify and cover more cases for bindMount.

Signed-off-by: Therese Nordqvist <therese.nordqvist@pelagicore.com>

### CODE REVIEW

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 1]
```diff
@@ -69,3 +69,43 @@ TEST_F(FileGatewayTest, TestActivateWithMinimalValidConf) {
     ASSERT_TRUE(gw->setConfig(jsonConfig));
     ASSERT_TRUE(gw->activate());
 }
+
+TEST_F(FileGatewayTest, MalformedPath) {
+// Mount a path which is malformed, e.g. too many "/"
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: I think the comment should either be indented along with the code inside the test, or located above the test if it's a comment for the whole test. Same goes for some other places below as well.

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 2]
```diff
@@ -69,3 +69,43 @@ TEST_F(FileGatewayTest, TestActivateWithMinimalValidConf) {
...
+    const std::string config =
+        "[{"
+        "  \"path-host\": \"//tmp//filename.txt\""
+        ", \"path-container\": \"///filename.txt\""
+        "}]";
+    loadConfig(config);
+
+    // Is this ok? Will the extra "/" will be removed?
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Is this a "todo" comment, or is it a comment for what's being asserted?

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 3]
```diff
@@ -69,3 +69,43 @@ TEST_F(FileGatewayTest, TestActivateWithMinimalValidConf) {
...
+    ASSERT_TRUE(gw->setConfig(jsonConfig));
+    ASSERT_TRUE(gw->activate());
+}
+
+TEST_F(FileGatewayTest, MountNonExisting) {
+// Mount a non-existing path
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Is this a comment for what the test does? Put it above the test in that case, and comment on what's being tested, e.g. "Mount a non-existing path and assert that ... "

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 4]
```diff
@@ -69,3 +69,43 @@ TEST_F(FileGatewayTest, TestActivateWithMinimalValidConf) {
...
+            ", \"path-container\": \"/tmp/thisFileDoesNotExist.txt\""
+        "}]";
+    loadConfig(config);
+
+    ASSERT_TRUE(gw->setConfig(jsonConfig));
+    ASSERT_FALSE(gw->activate());
+}
+
+TEST_F(FileGatewayTest, BadMount) {
+// Mount a non-existing path
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Copy pasted from above? Same comment as for above applies here I think.

**Review by breakreturn — CHANGES_REQUESTED** [review]

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp (hunk 2 above)
Comment (thenor) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Yes this is more or less on the level of a TODO. I saw the printout during the test and couldn't see if this path (with extra slashes) was what was being using for the bindmount command. I wanted to check up if we remove extra slashes (couldn't find it in our code, so perhaps it is the help functions we use). So it was a reminder to myself to check up on it.

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp (hunk 1 above)
Comment (thenor) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Yes, agreed. I wrote it in the beginning of the test case to specify what to test, then forgot to redo it as a test header/description.

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp (hunk 4 above)
Comment (thenor) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Yes, will redo all the comments and move them.

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 5]
```diff
@@ -37,27 +38,54 @@ class FileGatewayTest : public SoftwareContainerGatewayTest
 
         gw = std::unique_ptr<FileGateway>(new FileGateway(m_container));
 
-        // Create file
-        ASSERT_TRUE(writeToFile(FILE_PATH, FILE_CONTENT));
+        // Create files
+        ASSERT_EQ(mkdir(DIR_PATH.c_str(),775),0);
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: There should be spaces after the commas.

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 6]
```diff
@@ -37,27 +38,54 @@ class FileGatewayTest : public SoftwareContainerGatewayTest
...
+        ASSERT_TRUE(createFile(FILE_PATH, FILE_CONTENT));
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: I'm not a fan of testing things in the test setup. This also looks like it's actually testing the test setup code. If the test should break if this doesn't work I think some normal assert should work better.

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 7]
```diff
@@ -69,3 +97,117 @@ TEST_F(FileGatewayTest, TestActivateWithMinimalValidConf) {
...
+ */
+TEST_F(FileGatewayTest, MalformedPath) {
+    const std::string config =
+        "[{"
+        "  \"path-host\": \"//tmp//filename.txt\""
+        ", \"path-container\": \"///filename.txt\""
+        "}]";
+    loadConfig(config);
+
+    // TODO: Verify that this is not a false positive
```
Comment (breakreturn) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Should this be resolved or is it intentionally left?

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp (hunk 6 above)
Comment (thenor) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Ok. I will change it to use the regular c++ assert.

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp (hunk 7 above)
Comment (thenor) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: I forgot to remove it after I had tested this.

**Review by breakreturn — APPROVED** [review]

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 8]
```diff
@@ -37,27 +38,54 @@ class FileGatewayTest : public SoftwareContainerGatewayTest
...
         SoftwareContainerGatewayTest::TearDown();
     }
 
+    bool createFile(const std::string &path, const std::string &content)
+    {
+        bool ret = true;
+        std::ofstream out(path);
+        if (out.is_open()) {
+            out << content;
+            if (!out.good()) {
```
Comment (kursatkobya) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: just an option, you can get rid of ret variable just by returning good method to have concise function
```
std::ofstream out(path);
if (out.is_open()) {
    out << content;
    return out.good();
}

return false;
```

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 9]
```diff
@@ -37,27 +38,54 @@ class FileGatewayTest : public SoftwareContainerGatewayTest
...
+        return ret;
+    }
+
     const std::string FILE_CONTENT = "ahdkhqweuyreqiwenomndlaskmd";
     const std::string FILE_PATH = "/tmp/filename.txt";
+    const std::string DIR_PATH = "/tmp/testdir/";
+    const std::string LONG_FILE_PATH = DIR_PATH + "filename.txt";
+    const std::string FILE_CONTENT_2 = "bgsajbjkbjherkgbjhsbnjdfnjk";
+    const std::string FILE_W_SPACES = "example file name.txt";
+    const std::string FILE_PATH_W_SPACES = "/tmp/example file name.txt";
```
Comment (kursatkobya) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: this looks like filename with spaces to me.  Adding space after tmp would be more explanatory.
`/tmp /example file/name.txt`

**Review by kursatkobya — APPROVED** [review]
looks nice.
approved with small comments

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp (hunk 9 above)
Comment (thenor) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: But that would be a new directory: "tmp\ "
What I wanted was just "/tmp/"+FILE_W_SPACES, whic I could have written.

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp (hunk 8 above)
Comment (thenor) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Oh yes of course. Good catch.

Comment (thenor) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Hm well... I want to do the out.close also.

Comment (kursatkobya) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: I think ofstream destructor closes the file but sure

Comment (thenor) [reply libsoftwarecontainer/component-test/filegateway_componenttest.cpp]: Tbh I copied this code from common and didn't analyze it too much. Maybe you are right.

**Review by frznlogic — APPROVED** [review]

File: libsoftwarecontainer/component-test/filegateway_componenttest.cpp [hunk 10]
```diff
@@ -37,27 +38,48 @@ class FileGatewayTest : public SoftwareContainerGatewayTest
...
 
+    bool createFile(const std::string &path, const std::string &content)
+    {
+        std::ofstream out(path);
+        if (out.is_open()) {
+            out << content;
+            return out.good();
+        }
+        return false;
+    }
```
Comment (kursatkobya) [inline libsoftwarecontainer/component-test/filegateway_componenttest.cpp:67]: nice !!

**Pull request merged**